### PR TITLE
Clay v1 base

### PR DESCRIPTION
**Pull request by myscon**

### CODE REVIEW

**Review by Joao-L-S-Almeida — APPROVED** [review]

**Comment by Joao-L-S-Almeida** [comment]
The tests are failing due to an import for `vit_pytorch` and I understand that it should came from [here](https://pypi.org/project/vit-pytorch/).  Could it be replaced by some builtin module or should we add it to our requirements ?

**Comment by myscon** [comment]
Okay I removed vit-torch.  I seem to have omitted the sign off and that was the reason for the force.

**Comment by Joao-L-S-Almeida** [comment]
Some of the automatic tests are still failing. Maybe it could be interesting to run `pytest -s tests` locally and check if there any missing stuff.

**Comment by myscon** [comment]
Made some small adjustments to the test file. They pass locally.

**Comment by Joao-L-S-Almeida** [comment]
@myconv I saw that the problem was related to the device. I modified it to "cpu" (the GitHub runner was not configured with GPU) and the tests are passing. Please, check the PR #160, try to update your fork and we can proceed with the review.

File: terratorch/models/backbones/clay_v1/embedder.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+
+class Embedder(nn.Module):
+    default_out_indices = (0,)  # Single out_indices for simplicity
+
+    def __init__(self,
+                 img_size=256,
+                 num_frames=1,
+                 ckpt_path=None,
+                 device="cuda",
```
Comment (CarlosGomes98) [inline terratorch/models/backbones/clay_v1/embedder.py:34]: Device shouldn't be required to be passed. In pytorch lightning, `.to(device)` statements should not be used, as the responsibility to place modules in the correct device is passed to lightning.

File: terratorch/models/backbones/clay_v1/embedder.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        self.clay_encoder = (
+            EmbeddingEncoder(  # Default parameters for the Clay base model
+                img_size=img_size,
+                patch_size=8,
+                dim=768,
+                depth=12,
+                heads=12,
+                dim_head=64,
+                mlp_ratio=4.0,
+            ).to(device)
```
Comment (CarlosGomes98) [inline terratorch/models/backbones/clay_v1/embedder.py:58]: Same as above, such statements should be removed

Comment (myscon) [reply terratorch/models/backbones/clay_v1/embedder.py]: Makes sense. Those have been removed.

File: terratorch/models/backbones/clay_v1/modules.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+import torch.nn.functional as F
+from einops import rearrange, repeat
+from torch import nn, Tensor
+
+from terratorch.models.backbones.clay_v1.utils import posemb_sincos_1d, posemb_sincos_2d_with_gsd
+
+os.environ["TORCH_CUDNN_V8_API_DISABLED"] = "1"
+
+# central wavelengths of pretrained model
+WAVELENGTHS = {
```
Comment (CarlosGomes98) [inline terratorch/models/backbones/clay_v1/modules.py:13]: could this use `HLSBands` for the keys?

File: terratorch/models/backbones/clay_v1/modules.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        self.to_qkv = nn.Linear(dim, inner_dim * 3, bias=False)
+        self.to_out = nn.Linear(inner_dim, dim, bias=False)
+
+    def forward(self, x):
+        x = self.norm(x)
+
+        qkv = self.to_qkv(x).chunk(3, dim=-1)
+        q, k, v = map(lambda t: rearrange(
+            t, 'b n (h d) -> b h n d', h=self.heads), qkv)
+
```
Comment (CarlosGomes98) [inline terratorch/models/backbones/clay_v1/modules.py:60]:
```suggestion
        (rearrange(
            t, 'b n (h d) -> b h n d', h=self.heads) for t in qkv)
```
may be simpler

File: terratorch/models/backbones/clay_v1/modules.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+            datacube["gsd"],  # 1
+            datacube["waves"],  # [N]
+        )  # [B C H W]
+
+        B, C, H, W = cube.shape
+
+        patches, waves_encoded = self.to_patch_embed(
+            cube, waves
+        )  # [B L D] - patchify & create embeddings per patch
+        # TODO: Add time & latlon as encoding to patches
```
Comment (CarlosGomes98) [inline terratorch/models/backbones/clay_v1/modules.py:245]: is this still todo?

**Review by CarlosGomes98 — COMMENTED** [review]
Added some comments. This is quite a big task, so we may want to think about how we can split work somehow :)

**Comment by romeokienzler** [comment]
Thanks @myscon 

- [x] merged

Thanks so much @Joao-L-S-Almeida for your support and @paolo-fraccaro for kicking this off

**Comment by Joao-L-S-Almeida** [comment]
Thanks everyone.

**Pull request merged**